### DISCUSSION TITLE
Intended way of including python libraries (frozen) in the build...

### DISCUSSION THREAD

**Discussion by synco** [micropython/micropython#9381] [Libraries & Drivers] [answered]
Hello,
   I've compiled micropython for the ESP32 on custom hardware, and would like to select (include + remove) some of the frozen packages built into the image.
I have been copying the desired *.py files from the micropython/lib/micropython-lib/* into my /micropython/ports/esp32/modules/ directory, and it works perfectly well. (ie, they get compiled as *.mpy and included in the image) 
It just feels a bit wrong, and wondering what is this the proper/intended way?
And how do you remove unwanted drivers / libraries? 
Are the manifest files used for this?
Thanks in advance.

**Comment by jimmo** [comment, marked as answer]
> Are the manifest files used for this?

Yes. The idea is that you can write your own `manifest.py` (which lives outside the main repo, in your own repo ideally) and then pass that to the build with `make BOARD=... FROZEN_MANIFEST=...path...`.

Your custom manifest is free to use any of the base ones by `include()`-ing the port or board manifests.

The next step up is to define your own custom board definition (which sounds like might be useful to you if you're building a custom board) which lets you configure much more (via a custom mpconfigboard.cmake and mpconfigboard.h). These also can live in your own repo. See https://github.com/micropython/micropython-example-boards

(This is a bit of an FAQ, and we really need to write more documentation on this process...).

**Reply by puppet13th** [reply]
@synco you can add subdirectory on '/micropython/ports/esp32/modules/' too:

i have compiled mpy to include aioble library.i added these files to modules:
```
/micropython/ports/esp32/modules/aioble/central.py
/micropython/ports/esp32/modules/aioble/client.py
/micropython/ports/esp32/modules/aioble/core.py
/micropython/ports/esp32/modules/aioble/device.py
/micropython/ports/esp32/modules/aioble/l2cap.py
/micropython/ports/esp32/modules/aioble/peripheral.py
/micropython/ports/esp32/modules/aioble/security.py
/micropython/ports/esp32/modules/aioble/server.py
/micropython/ports/esp32/modules/aioble/__init__.py
```
help('modules') output:
```
aioble/__init__   dht               uasyncio/stream   usocket
aioble/central    ds18x20           ubinascii         ussl
aioble/client     esp               ubluetooth        ustruct
aioble/core       esp32             ucollections      usys
aioble/device     flashbdev         ucryptolib        utime
aioble/l2cap      framebuf          uctypes           utimeq
aioble/peripheral gc                uerrno            uwebsocket
aioble/security   inisetup          uftpd             uzlib
aioble/server     math              uhashlib          webrepl
```
